### ISSUE TITLE
Add support for new proxy `action` names used by v2 proxy

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#4383]
Historically, the proxy has used the `Google_Proxy::ACTION_SETUP` name for the `action` query parameter that it uses to pass certain data back to Site Kit as part of some intermediate redirects during the plugin setup flow. This has been somewhat problematic since it uses a single `action` for multiple interactions, and it's even worse because the same action is also used by the plugin internally to redirect the user from an "internal" setup request to the proxy setup.

Since we are now making a clean break by migrating to a `v2/site-management/setup/` action, this is a great point to resolve this mess. The proxy will now, whenever the new route is used, send this data through distinct `action` query parameters, which this issue aims at adding support for.

**The work in this issue is closely related to #2044 and should preferably be worked on in the same sprint.**

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* Instead of handling all intermediate proxy redirects via the `googlesitekit_proxy_setup` (`Google_Proxy::ACTION_SETUP`) action (which is also used by the plugin internally for another purpose), this should now be taken care of in a more granular way _if_ the `serviceSetupV2` feature flag is enabled:
    * For the request to verify the site (includes verification token), the proxy now sends a new `googlesitekit_proxy_verify` action.
    * For the request to purely exchange the site code for the actual site credentials, the proxy now sends a new `googlesitekit_proxy_exchange_site_code` action.
* The logic that is currently in place to handle both of these interactions can pretty much remain the same. No parameters sent are changed, it's purely the `action` query parameter that has the names updated as above. However, having these two distinct action allows us to clean up the code quite a bit and split the handlers more appropriately - up to the IB to solve this nicely. :)
* The only other behavioral change that needs to be covered here is the following: Both actions should check for an optional `step` query parameter sent from the proxy. If that is received, its value should be added to the redirect URL to send back to as well. (This is typically going to be `verification` for the `googlesitekit_proxy_verify` action and `delegation_consent` for the `googlesitekit_proxy_exchange_site_code` action.)
* [This commit in the proof-of-concept branch](https://github.com/google/site-kit-wp/commit/ad097bf0269eb270d9a805e4266d1ee24a28abb2) pretty much contains the changes needed here, however in a somewhat quick and hacky implementation - looking at this can still serve as inspiration or to better clarify the requirements.
* The existing handlers for the `googlesitekit_proxy_setup` still need to be kept/maintained for when the feature flag is not enabled.

## Implementation Brief

We should extract setup-related functionality from the main `Authentication` class which is already quite large. This will help us to better reason about the differences between setup v1 and v2 while minimizing the use of feature flags.

Additionally, we should decouple the `ACTION_SETUP` action from handling setup functions _and_ beginning the setup/site registration flow.

The current action handler to start setup should be updated to use a new `ACTION_SETUP_START` while the current `ACTION_SETUP` is only used for v1 setup flow requests.

**`Google_Proxy`**

* Add new class `ACTION_` constants for the new actions Site Kit will respond to
    - `ACTION_SETUP_START` : `googlesitekit_proxy_setup_start`
    - `ACTION_VERIFY` : `googlesitekit_proxy_verify`
    - `ACTION_EXCHANGE_SITE_CODE` : `googlesitekit_proxy_exchange_site_code`

**Create a new `Authentication\Setup` base class for common setup methods**

The class should accept/require `Context`, `User_Options` and `Authentication` as constructor dependencies. All other classes (e.g. `Google_Proxy`) are available via `Authentication`.

This should be an abstract class, with abstract `register` method, in addition to common methods for sub-classes, such as the following. Most can be moved over from `Authentication` with little to no changes
* `verify_nonce( $action, $nonce )`
* `handle_verification()` to use a new action internally for handling the verification and passing through the token and type/method values via context->input  
    ```php
    do_action( 'googlesitekit_verify_site_ownership', $verification_token, $verification_type );
    ```
* `handle_site_code( $code, $site_code )`
* `redirect_to_proxy( $code )`
    Should be updated to include a given `step` via the current request's query parameter in the redirect URL back to the proxy if present (this is only really relevant for v2 but should be safe to include generally, alternatively this detail could be implemented in the `V2` class specifically)

**Create a new `Authentication\Setup_V1` class that extends `Setup`**

This class will handle setup behavior for v1 of the setup flow only.

* `register` should add an hook on `'admin_action_' . Google_Proxy::ACTION_SETUP_START` to a `handle_action_setup_start` callback
    - Essentially handles `redirect_to_setup_flow` + `redirect_to_proxy` as today but only for the v1 case (no feature flag logic)
* `register` should add an hook on `'admin_action_' . Google_Proxy::ACTION_SETUP` to a single `handle_action_setup_start` callback for handling the setup flow actions
    - nonce verification for `Google_Proxy::ACTION_SETUP`
    - `handle_verification()`
    - `handle_site_code( $code, $site_code )`
    - `redirect_to_proxy( $code )`

**Create a new `Authentication\Setup_V2` class that extends `Setup`**

This class will handle setup behavior for v2 of the setup flow only.

* `register` should add an hook on `'admin_action_' . Google_Proxy::ACTION_SETUP_START` to a `handle_action_setup_start` callback
    - Essentially handles `redirect_to_setup_flow` + `redirect_to_proxy` as today but only for the v2 case (no feature flag logic)
* `register` should add an hook on `'admin_action_' . Google_Proxy::ACTION_VERIFY` to a single `handle_action_verify` callback; essentially
    - nonce verification for `Google_Proxy::ACTION_VERIFY`
    - `handle_verification()`
    - `redirect_to_proxy( $code )`
* `register` should add an hook on `'admin_action_' . Google_Proxy::ACTION_EXCHANGE_SITE_CODE` to a single `handle_action_exchange_site_code` callback; essentially
    - nonce verification for `Google_Proxy::ACTION_EXCHANGE_SITE_CODE`
    - `handle_site_code( $code, $site_code )`
    - `redirect_to_proxy( $code )`

**`Authentication`**

* Remove all `ACTION_SETUP*` hooks and related methods that are no longer used
* Update the action provided in `get_proxy_setup_url` to use `ACTION_SETUP_START`

**`Site_Verification`**

* Update the action its `handle_verification_token` method is hooked to the new `googlesitekit_verify_site_ownership` action used in both versions of the setup flow now

**`Plugin`**

* Initialize and register `Setup_V1` or `Setup_V2` (not both) based on the active state of the `serviceSetupV2` flag respectively (v2 if flag enabled otherwise v1)

### Test Coverage

* Add test cases for `Setup_V1` and `Setup_V2` classes
* Update `Site_VerificationTest` to use the new dedicated action instead of the `SETUP` action
* Update `AuthenticationTest` to remove test coverage for functionality moved to `Setup*` classes

## QA Brief

* Test the current setup flow **without** the `serviceSetupV2` feature flag enabled
    * Make sure to test this on a fresh site (never used with the proxy before) to ensure the verification mechanic works
    * Make sure to also test this on a site which has already been verified

`QA:Eng` – secondary review of implementation to make sure ACs are met and no regressions are introduced that might not be caught in normal QA.

## Changelog entry

* N/A

**Comment by felixarntz** [comment]
cc @aaemnnosttv for visibility

@FlicHollis This is another issue for `serviceSetupV2` which would be good to tackle in Sprint 64. I'd say let's prioritize this and #2044 for Sprint 64 - #2045 can then be in either Sprint 64 or Sprint 65.

**Comment by aaemnnosttv** [comment]
> * When redirecting back to the proxy after having received the `googlesitekit_proxy_verify` action, the redirect URL should have the `step=verification` query parameter set.
> * When redirecting back to the proxy after having received the `googlesitekit_proxy_exchange_site_code` action, the redirect URL should have the `step=delegation_consent` query parameter set.

@felixarntz IB added with one question: should the proxy send a `redirect_url` parameter rather than rely on the plugin to set the proper `step` to return with? The plugin could set/alter params as necessary, but it seems maybe the return URL should come from the proxy itself. What do you think?

**Comment by felixarntz** [comment]
@aaemnnosttv Good idea! I think it's more straightforward though to just pass the `step` from the proxy as a query parameter instead of the entire URL. Then the plugin adds the same value as `step` parameter to the URL to redirect back to, kind of similar how it does it with `code` and `site_code` for example. I've updated the ACs with that approach.

That part now needs to be updated in the IB, other than that it mostly looks good. A few other points:

> Update `Google_Proxy::get_supports` to conditionally include a new `service_setup_v2` feature if the `serviceSetupV2` feature flag is enabled

Why is this necessary? I don't think the proxy needs that, since the entire setup route is entirely different there (see #2044), and only the new one uses the new behavior.

> When the `serviceSetupV2` feature is enabled, hook alternate callbacks in place of the existing inline function which handles the site code and redirects back

Related to #4376, I think it would be worth simplifying this code here. Specifically I'm thinking maybe we could consolidate the hook callbacks into fewer so that the code is easier to follow sequentially, rather than having to figure out what callback fires when and mentally put together the pieces. There's still the part that we'd like to keep the site verification bit within the `Site_Verification` module class, but we can probably figure it out some way.
* At a minimum it would be good to put the logic where it redirects back to the proxy into the same method which is currently the "main handler" for the action (we typically have three methods right now: nonce verification, main handler, redirect back). This allows us to get rid of the filter later, which would be a nice clean-up - relying purely on method parameters to specify query parameters is more explicit and easier to follow (see https://github.com/google/site-kit-wp/issues/4376#issuecomment-975906978 for background).
* Alternatively, we could even use a single handler method per each action, that would make it really easy to understand what's going on. For the part of the code that we have in `Site_Verification`, I wouldn't be opposed to introducing a dedicated action hook that is specifically for that part, e.g. a `do_action( 'googlesitekit_verify_site_ownership', $verification_token, $verification_method )`.

I think the second point above would be the cleanest implementation. We can still use reusable methods for the common parts of the logic between the two handlers, but having only one handler per action would make this a lot easier to understand IMO. What do you think?

**Comment by aaemnnosttv** [comment]
> > Update `Google_Proxy::get_supports` to conditionally include a new `service_setup_v2` feature if the `serviceSetupV2` feature flag is enabled
> 
> Why is this necessary? I don't think the proxy needs that, since the entire setup route is entirely different there (see #2044), and only the new one uses the new behavior.

@felixarntz yes, I see now that it uses the separate URL 👍 

Agreed on the other points, I'll update the IB to resolve some of this technical debt.

**Comment by aaemnnosttv** [comment]
@felixarntz I've updated the IB here with something I think should work well. If you agree, this would probably be good for me to pick up since I have some of this sketched out locally already.

**Comment by felixarntz** [comment]
@aaemnnosttv Thanks, the suggested approach sounds great - IB ✅ 

I'll assign this to you already in EB.

**Comment by felixarntz** [comment]
For reference: As discussed in our sync call, this is now a blocker for #4376 so it is now part of the current sprint in favor of the other one which can be worked on once this is merged, preferably in the following sprint.

**Comment by aaemnnosttv** [comment]
@felixarntz is it to be expected that the setup flow should work with both v1+v2 once merged? I assume yes because it works now but just wanted to double-check for the QAB.

**Comment by aaemnnosttv** [comment]
@felixarntz should this be QA:Eng then or should we limit what is QA'd for V2 for now?

**Comment by felixarntz** [comment]
@aaemnnosttv It should definitely include regular QA to ensure it doesn't break V1. Since it doesn't really allow QAing V2 though, I'd say either we don't do that here at all, or (preferably) we have a two-step QA here, the second of which would be `QA: Eng`. Maybe @wpdarren can start with V1 QA and then add the label for the eng part about V2?

**Comment by wpdarren** [comment]
@felixarntz sounds good to me. We've done this before on a few other tickets.  Once I have completed my testing I will add the `QA:Eng` label.

**Comment by eclarke1** [comment]
@felixarntz @aaemnnosttv could I just chase up getting this ticket moving please, as it's blocking a few tickets in IB assigned to this sprint. What are the next steps here please? Thank you. cc @FlicHollis

**Comment by aaemnnosttv** [comment]
@eclarke1 it's with @felixarntz awaiting code review 👍

**Comment by wpdarren** [comment]
### QA Update: ⚠️ 

@aaemnnosttv I an observation that I pretty sure has not occurred previously.  

I am testing this on a new site and also a new Google account. 

When I authorise my new Google account for the first time this message appeared on the next screen. I've disconnected, reset Site Kit and the message no longer appears when connecting Site Kit  Anything to be concerned about?

`[amp-form] Form submission failed: Error: XHR Failed fetching`

![image](https://user-images.githubusercontent.com/73545194/145403228-8d4b3419-73df-4f93-b0d1-561a562dcb6d.png)

I was able to connect Site Kit, add site and was successfully redirected to the dashboard.

**Comment by aaemnnosttv** [comment]
@wpdarren that doesn't look related to the plugin since you're on step 2 there which is a proxy-side only thing. Does this happen every time for you on that step? There could be something that needs attention but it's hard to say without seeing the full URL. 

I just tried it with a new site and didn't see the same thing so it doesn't seem like a problem but maybe something for @felixarntz to confirm.

**Comment by wpdarren** [comment]
@aaemnnosttv I will try recreate and get you the full URL and error.

**Comment by wpdarren** [comment]
@aaemnnosttv I didn't receive that message this time around. Created new site and google account.

### QA Update: ✅ 

- Tested the current setup flow **without** the `serviceSetupV2` feature flag enabled.
- Went through the set up on a fresh site and also new Google account.
- No issues occurred while setting up Site Kit or while connecting the modules. 
- Went through the set up on a site which had been verified. 
- No issues occurred while setting up Site Kit or while connecting the modules.

Added the `QA:Eng` label on for the second part of testing.

**Comment by tofumatt** [comment]
QA ✅  from me.

I also ran through the regular QA steps, but then looked through the code changes and didn't see any obvious places for regressions, so I think we're good 👍🏻
